instrumentation_sql.go

```diff
@@ -0,0 +1,413 @@
+		stmt, err = c.PrepareContext(ctx, query)
+	} else {
+		stmt, err = conn.Prepare(query)
+	}
+
+	if err != nil {
+		return stmt, err
+	}
+
+	if stmt, ok := stmt.(*wrappedSQLStmt); ok {
+		return stmt, nil
+	}
+
+	return &wrappedSQLStmt{
+		Stmt:        stmt,
+		connDetails: conn.details,
+		query:       query,
+		sensor:      conn.sensor,
+	}, nil
+}
+
+func (conn *wrappedSQLConn) ExecContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Result, error) {
+	sp := startSQLSpan(ctx, conn.details, query, conn.sensor)
+	defer sp.Finish()
+
+	if c, ok := conn.Conn.(driver.ExecerContext); ok {
+		res, err := c.ExecContext(ctx, query, args)
+		if err != nil && err != driver.ErrSkip {
+			sp.LogFields(otlog.Error(err))
+		}
+
+		return res, err
+	}
+
+	if c, ok := conn.Conn.(driver.Execer); ok { //nolint:staticcheck
+		values, err := sqlNamedValuesToValues(args)
+		if err != nil {
+			return nil, err
+		}
+
+		select {
+		default:
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+
+		res, err := c.Exec(query, values)
+		if err != nil && err != driver.ErrSkip {
+			sp.LogFields(otlog.Error(err))
+		}
+
+		return res, err
+	}
+
+	return nil, driver.ErrSkip
+}
+
+type wrappedSQLStmt struct {
+	driver.Stmt
+
+	connDetails dbConnDetails
+	query       string
+	sensor      *Sensor
+}
+
+func (stmt *wrappedSQLStmt) ExecContext(ctx context.Context, args []driver.NamedValue) (driver.Result, error) {
+	sp := startSQLSpan(ctx, stmt.connDetails, stmt.query, stmt.sensor)
+	defer sp.Finish()
+
+	if s, ok := stmt.Stmt.(driver.StmtExecContext); ok {
+		res, err := s.ExecContext(ctx, args)
+		if err != nil && err != driver.ErrSkip {
+			sp.LogFields(otlog.Error(err))
+		}
+
+		return res, err
+	}
+
+	values, err := sqlNamedValuesToValues(args)
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	default:
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+
+	res, err := stmt.Exec(values) //nolint:staticcheck
+	if err != nil && err != driver.ErrSkip {
+		sp.LogFields(otlog.Error(err))
+	}
+
+	return res, err
+}
+
+func (stmt *wrappedSQLStmt) QueryContext(ctx context.Context, args []driver.NamedValue) (driver.Rows, error) {
+	sp := startSQLSpan(ctx, stmt.connDetails, stmt.query, stmt.sensor)
+	defer sp.Finish()
+
+	if s, ok := stmt.Stmt.(driver.StmtQueryContext); ok {
+		res, err := s.QueryContext(ctx, args)
+		if err != nil && err != driver.ErrSkip {
+			sp.LogFields(otlog.Error(err))
+		}
+
+		return res, err
+	}
+
+	values, err := sqlNamedValuesToValues(args)
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	default:
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+
+	res, err := stmt.Stmt.Query(values) //nolint:staticcheck
+	if err != nil && err != driver.ErrSkip {
+		sp.LogFields(otlog.Error(err))
+	}
+
+	return res, err
+}
+
+func startSQLSpan(ctx context.Context, conn dbConnDetails, query string, sensor *Sensor) ot.Span {
+	tags := ot.Tags{
+		string(ext.DBType):      "sql",
```

Comment:
@andrewslotin I would be nice to set the database type whenever it's known. For example if the mysql strategy was used to parse the connection string, we could send "mysql".

Reply:
The OpenTracing spec [prescribes](https://opentracing.io/specification/conventions/) the use of `sql` for any SQL database:

> Database type. For any SQL database, "sql". For others, the lower-case database category, e.g. "cassandra", "hbase", or "redis".

Reply:
right so "mysql" if we know it's mysql, etc. and then "sql" otherwise.

Reply:
Not exactly. If this is an SQL database, then "sql". Otherwise put the lower-case database category, e.g. "mongo" or "cassandra".

Reply:
sorry read too fast, again... thanks for being patient ;)